### PR TITLE
Update: Add support for x_forwarded_for headers in apaches access logs

### PR DESCRIPTION
**Pull request by redcinelli**
## What does this PR do?

Tldr;
Enable support for apache logs with the x_forwarded_for header.

Heavily inspired by the work done in [this PR](https://github.com/elastic/beats/pull/4417)
Addressing [this Enhancement Request](https://github.com/elastic/enhancements/issues/14402).

`Grok` pattern has been updated to match logs starting with a list of IP, storing it in `apache.access.remote_ip`.
This pattern is heavily inspired by the [nginx integration](https://github.com/elastic/integrations/blob/main/packages/nginx/data_stream/access/elasticsearch/ingest_pipeline/default.yml).

A painless scrip extract the first public ip, and store the result in `source.address`.

`network.forwarded_ip` field is being filled by `source.ip`, as it seems to be the perfect fit.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## Author's Checklist

- [x] Do we want to support `x_forward_for` ?
- [x] Lots of noise in the test files, let's double check it is really backward compatible.
- [x] Simplify the Grok pattern, verify the Grok pattern
- [x] I am allowed to modify the `httpd.conf` ?

## How to test this PR locally

Following the steps in the documentation
```
elastic-package stack down && elastic-package build && elastic-package stack up -d -v --services=elasticsearch && eval "$(elastic-package stack shellinit)" && elastic-package test pipeline --generate
```

## Related issues

None

## Screenshots

None

### CODE REVIEW

**Comment by redcinelli** [comment]
I believe my pr to be good for review 🤞 

At least open for discussion.

Btw seeing the whole mess the `.json` files are should I first do a pr to "update" the test then rebase this pr ?

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 1]
```diff
@@ -30,6 +30,65 @@ processors:
...
         HTTP/%{NUMBER:http.version}" (-|%{NUMBER:http.response.body.bytes:long})
       ignore_missing: true
+      pattern_definitions:
+        ADDRESS_LIST: (?:%{IP}|%{WORD})("?,?\s*(?:%{IP}|%{WORD}))*
+  - split:
+      field: apache.access.remote_ip_list
+      separator: '"?,?\s+'
+      ignore_missing: true
+  - script:
+      if: ctx.apache?.access?.remote_ip_list != null && ctx.apache.access.remote_ip_list.length > 0
```
[the remark this answers is not in the document]
Comment (redcinelli) [reply packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml:40]: done in #[6489491](https://github.com/elastic/integrations/pull/3251/commits/6489491ec47df082476c5cef15adc70dc98a4289)

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 2]
```diff
@@ -30,6 +30,65 @@ processors:
...
+      lang: painless
+      source: >-
+        boolean isPrivate(def dot, def ip) {
```
[the remark this answers is not in the document]
Comment (redcinelli) [reply packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml]: Indeed it is very much cleaner.

File: packages/apache/data_stream/access/fields/fields.yml [hunk 3]
```diff
@@ -9,3 +9,8 @@
       type: keyword
       description: |
         SSL cipher name.
+        - name: nginx.access
+    - name: remote_ip_list
+      type: array
```
[the remark this answers is not in the document]
Comment (redcinelli) [reply packages/apache/data_stream/access/fields/fields.yml]: Done

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (redcinelli) [reply packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml]: I have been using this https://en.wikipedia.org/wiki/Private_network to create the filter.

Do you think we should integrate the `Dedicated space for carrier-grade NAT deployment` and `Private IPv6 addresses`

**Comment by redcinelli** [comment]
@andrewkroh I believe it is worth taking a look again ^^

Thanks a lot for helping me.

**Review by andrewkroh — COMMENTED** [review]
I was reviewing this from technical perspective on the code/config in the pipeline. I didn't really consider the big picture with respect to ECS mappings, the field meanings, and whether this is a breaking change.

Taking a step back and looking at the approach here, this looks like a change in behavior w.r.t. to how `source` is populated. IMO the `source.address` should be the IP (or hostname) where traffic came from (the last hop). That is `%{c}h` or `%h` or (barring any changes from things like mod_remoteip) in mod_log_config format. This would be the same source address observed if we looked at a packet capture.

The ECS `network.forwarded_ip` would be populated only when `X-Forwarded-For` is used because this field is only used when the source address represents a proxy. I think the first address in the X-Forwarded-For list would always be used to fill `network.forwarded_ip`.

I think the new `apache.access.remote_ip_list` field should be the x-forwarded-for list plus the remote hostname (%h value). This would mean that all the hops from the client through the last proxy are accounted for in the list. And given the reference info you pointed me to, perhaps the name of this field is better as `apache.access.remote_addresses` (data type `keyword`) so that we can take the raw values without caring if they are IPs.

In order to have both `%h` and `%{X-Forwarded-For}i` available in the log I think we should prescribe formats in the docs that it will accept. We want to continue accepting the current format as is and we also want to allow X-Forwarded-For to be added by users. So perhaps we should allow another header to be tacked on to the end like:

`%h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\" X-Forwarded-For=\"%{X-Forwarded-For}i\"`

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 4]
```diff
@@ -30,6 +30,50 @@ processors:
...
+              ctx.source.address = item;
+              found = true;
+              break;
+            }
+          }
+          if (!found) {
+            ctx.source.address = ctx.apache.access.remote_ip_list[0];
+          }
+        }
+        catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (redcinelli) [reply packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml:74]: Agreed, but this should not happened, and if it does. I would rather we find it early and create a fix, than failing silently.

**Comment by redcinelli** [comment]
Hi Andrew, so 2 new commits:

- [e3f15b9](https://github.com/elastic/integrations/pull/3251/commits/e3f15b912c7fff0222e05a00faa3e1fd7e9186e3) to address your suggestion 

- [4abfdbb](https://github.com/elastic/integrations/pull/3251/commits/4abfdbb062ebe8107da80f669b913f54243144a1) to try to address the bigger picture

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 5]
```diff
@@ -16,20 +16,69 @@ processors:
...
+        try {
+          if (ctx.source == null) {
+            Map map = new HashMap();
+            ctx.put("source", map);
+          }
+
+          def found = false;
+          for (def item : ctx.apache.access.remote_addresses) {
+            if (!isPrivateCIDR(item)) {
+              ctx.source.address = item;
```
Comment (andrewkroh) [inline packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml:70]: I don't think the `source.address` should be modified from what we got directly from the apache log (`%h`). I would probably put this into its own field withing the `apache.access` namespace. Or perhaps the `client.address` and `client.ip` are appropriate?

Comment (redcinelli) [reply packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml:70]: Hmmm, my bad.
I misunderstood you.

I'll set `client.ip`

**Review by andrewkroh — COMMENTED** [review]
Overall LGTM, but I think there is one thing missing -- docs. Users need to know how to configure there HTTPD service to output one of the formats the integration expects. Can you take a look at the readme (should be in packages/apache/_dev/...) and see if you can add a blurb about the accepted formats.

File: packages/apache/_dev/build/docs/README.md [hunk 6]
```diff
@@ -16,6 +16,24 @@ Access logs collects the Apache access logs.
...
+  - Defined in apache `LogFormat` by : 
+    >```%h %l %u %t \"%r\" %>s %b```
+  - Example:
+    > `127.0.0.1 user-identifier frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326`
+- Combined Log Format
+  - Defined in apache `LogFormat` by:
+    >```%h %l %u %t \"%r\" %>s %b \"%{Referer}i\" \"%{User-Agent}i\"```
+  - Example:
+    >```127.0.0.1 user-identifier frank [10/Oct/2000:13:55:36 -0700] "GET /apache_pb.gif HTTP/1.0" 200 2326 "http://datawarehouse.us.oracle.com/datamining/contents.htm" "Mozilla/4.7 [en] (WinNT; I)"```
+- Combined Log Format + x_forwarded_for header
```
Comment (andrewkroh) [inline packages/apache/_dev/build/docs/README.md:31]:
```suggestion
- Combined Log Format + X-Forwarded-For header
```

**Review by andrewkroh — COMMENTED** [review]
The doc additions look good. The package needs to be build to update the generated readme file (e.g. `elastic-package build`). Then there is a conflict with the changelog and version that need resolved.

**Review by andrewkroh — APPROVED** [review]
LGTM.

This will require a review from @elastic/obs-service-integrations.

**Comment by redcinelli** [comment]
@elastic/obs-service-integrations Hi guys ? have you had the time to look at this PR ?

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -16,20 +16,69 @@ processors:
...
+      pattern_definitions:
+        ADDRESS_LIST: (%{IP})("?,?\s*(%{IP}))*
+  - split:
+      field: apache.access.remote_addresses
+      separator: '"?,\s*'
+      ignore_missing: true
+  - set:
+      field: network.forwarded_ip
+      value: "{{{apache.access.remote_addresses.0}}}"
+      if: ctx.apache?.access?.remote_addresses != null && ctx.apache.access.remote_addresses.length > 0
```
Comment (ishleenk17) [inline packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml:38]: Do we need both these checks ?
Not null will ensure length greater than 0 ?

File: packages/apache/data_stream/access/fields/ecs.yml [hunk 8]
```diff
@@ -30,6 +32,8 @@
   name: log.level
 - external: ecs
   name: message
+- external: ecs
+  name: network.forwarded_ip
```
Comment (ishleenk17) [inline packages/apache/data_stream/access/fields/ecs.yml:36]: We are setting the value of network.forwarded_ip in the ingest pipeline. 
Do we need to use the ecs version of this here ?

File: packages/apache/data_stream/access/_dev/test/pipeline/test-access-basic.log [hunk 9]
```diff
@@ -4,4 +4,7 @@
 172.17.0.1 - - [29/May/2017:19:02:48 +0000] "GET /stringpatch HTTP/1.1" 404 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
 monitoring-server - - [29/May/2017:19:02:48 +0000] "GET /status HTTP/1.1" 200 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
 127.0.0.1 - - [02/Feb/2019:05:38:45 +0100] "-" 408 152 "-" "-"
-monitoring-server - - [29/May/2017:19:02:48 +0000] "GET /A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4 HTTP/1.1" 200 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" "-"
\ No newline at end of file
+monitoring-server - - [29/May/2017:19:02:48 +0000] "GET /A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4 HTTP/1.1" 200 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" X-Forwarded-For="-"
+89.160.20.112 - - [29/May/2017:19:02:48 +0000] "GET /A%20Beka%20G1%20Howe/029_AND_30/15%20reading%20elephants.mp4 HTTP/1.1" 200 612 "-" "Mozilla/5.0 (Windows NT 6.1; rv:15.0) Gecko/20120716 Firefox/15.0a2" X-Forwarded-For="10.0.0.2,10.0.0.1"
```
Comment (ishleenk17) [inline packages/apache/data_stream/access/_dev/test/pipeline/test-access-basic.log:8]: The x-forwarded is present in the end only for these formats of logs and not the ones which matches the other grok patterns ?

Comment (redcinelli) [reply packages/apache/data_stream/access/_dev/test/pipeline/test-access-basic.log:8]: The header x-forwarded-for is only present for access logs.

Also, it is optional, so we need to have test with the header and without

File: packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml (hunk 7 above)
Comment (redcinelli) [reply packages/apache/data_stream/access/elasticsearch/ingest_pipeline/default.yml:38]: I believe we need to check both. `remote_addresses` may not be null but can have for value `""` an empty string of length 0

File: packages/apache/data_stream/access/fields/ecs.yml (hunk 8 above)
Comment (redcinelli) [reply packages/apache/data_stream/access/fields/ecs.yml:36]: I believe we have to.

If I do not I get this issue

```
FAILURE DETAILS:
apache/access test-access-basic.log:
[0] field "network.forwarded_ip" is undefined
```

while running this command `elastic-package build && elastic-package stack up -d -v --services=elasticsearch && eval "$(elastic-package stack shellinit)" && elastic-package test pipeline --generate`

Comment (ishleenk17) [reply packages/apache/data_stream/access/fields/ecs.yml:36]: Let's move it to fields.yml then as assignment is happening under ingest pipeline

Comment (redcinelli) [reply packages/apache/data_stream/access/fields/ecs.yml]: done in #ddaed22

Comment (redcinelli) [reply packages/apache/data_stream/access/fields/ecs.yml:36]: We decided to undo ddaed22

**Review by ishleenk17 — COMMENTED** [review]
Looks good !

**Review by ishleenk17 — APPROVED** [review]
